File: listsofwikis/checkalive.pl
#!/usr/bin/perl

# Name: checkalive.pl v1.2
# Description: This script will go thru a list of URLs & determine 
# if they are online & if they are Mediawiki wikis. It should work
# with: "/index.php/Main_Page", "index.php", "api.php" and even pages
# such as: "/wiki/Pagina_principale".
#
# Created: 12/14/2013
# Most recently updated: 01/26/2014 (It's a work-in-progress...)
# Copyright (c) 2013-2014 by Scott D. Boyd - scottdb56@gmail.com
# ====================================================================
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
# ====================================================================
#
# NOTE: The following three Perl modules need to be installed on your computer.
# Search for them on cpan.org or use your Linux distro's package manager.
use LWP::Simple;
use LWP::UserAgent;
use Crypt::SSLeay;
my $slp=2; # You can change this number for seconds to sleep between requests (currently 2 seconds)
my $urllist="my-URL-list.txt";
my $alivelist="alive-wikis.txt";
my $deadlist="dead-wikis.txt";
my $pwrdby1="Powered by MediaWiki";
my $pwrdby2="poweredby_mediawiki";
my $genmw="meta name=\"generator\" content=\"MediaWiki"
my $mwapi="MediaWiki API documentation page";
my $lw=0; my $dw=0;
my $a=1; my $b=0; my $c=0;
my $flag=0;
my $ua = LWP::UserAgent->new;
$ua->agent("Mozilla/5.0");			# use this user-agent to get into wikis that block spiders & robots
$ua->timeout(30);
$ua->show_progress(1);

# Here's where most of the work takes place:
open (MYURLLIST,"<$urllist")
  or die "Cannot open the URL-list file: $!";
open (ALIVEFILE,">$alivelist");
open (DEADFILE,">$deadlist");
while (<MYURLLIST>) {
  if ((/\#(.*?)/) || (/^\s*$/)) {		# check to see if line is a comment or a blank line
    next;					# if so - skip it
  } else {
    $url=$_;					# assign the current line to $url
    chomp $url;
    $req = HTTP::Request->new(GET => $url);  	#				 --|
    $req->header('Accept' => 'text/html');	#				   |-- some of these lines
    $res = $ua->request($req);			# send request 			   |-- were adapted from
    if ($res->is_success) {			# if the URL still exists	   |-- lwpcook.pod
       print "Got it! ";			#				   |
       $doc=$res->content;			#				   |
       &ParsePage;				# go to "ParsePage" sub-routine	   |
    } else {					#				   |
      $errormsg=$res->status_line;		#				 --|
      if (substr($errormsg,0,3) eq "500") {		# if response-code 500
         print DEADFILE "$url\n"; $dw++;
      } elsif  (substr($errormsg,0,3) eq "401") {	# if Unauthorized (code 401)
         print DEADFILE "$url\n"; $dw++;
      } elsif  (substr($errormsg,0,3) eq "403") {	# if forbidden (code 403)
         print DEADFILE "$url is alive but access is denied.\n"; $dw++;
      } elsif  (substr($errormsg,0,3) eq "404") {	# if URL is dead (code 404)
         print DEADFILE "$url\n"; $dw++;
      } elsif  (substr($errormsg,0,3) eq "406") {	# if Not Acceptable (code 406)
         print DEADFILE "$url\n"; $dw++;
      }
    }
    $c++; $b=$c/10; 
    if ($b==$a) { 
       print "Checked $c URLs -- ";		# print the progress every 10 URLs
       $a++;
    } 
    if ($slp > 0) {
       print "Pausing for $slp seconds...\n\n"; sleep $slp;
    } else { 					# don't pause - go on to the next URL
    }
  }
}
close DEADFILE; close ALIVEFILE; close MYURLLIST;
print "\nFinished! I found $lw live wikis and $dw dead or non-Mediawiki wikis.\n";

# Here's the sub-routine
# ======================
sub ParsePage  {
   print "Parsing the document... ";
   if (($doc=~/$pwrdby1/i) || ($doc=~/$pwrdby2/i)) {	# if the page contains: "Powered by MediaWiki"
      print "It's alive and powered by Mediawiki\n";	# or: "poweredby_mediawiki"
      print ALIVEFILE "$url\n";				# then it's a MediaWiki wiki
      $flag=1;$lw++;
   } elsif ($doc=~/$genmw/i) {				# if the content generator is MediaWiki
      print "It's alive and powered by Mediawiki\n";	# then it's a MediaWiki wiki
      print ALIVEFILE "$url\n";
      $flag=1;$lw++;
   } elsif ($doc=~/$mwapi/i) {				# if the api.php contains: "MediaWiki API documentation page"
      print "It's alive and powered by Mediawiki\n";	# then it's a MediaWiki wiki
      print ALIVEFILE "$url\n";
      $flag=1;$lw++;
   }
   unless ($flag) {
     print "It's alive but NOT powered by MediaWiki\n";
     print DEADFILE "$url is alive but NOT powered by Mediawiki\n"; $dw++;
   }
   $flag=0;
}

File: resources/listsofwikis/mediawiki/checkalive.py
#!/usr/bin/env python3

# Copyright (C) 2011-2012 WikiTeam
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# Script to check if a list of wikis are alive or dead

import re
import threading
import time
import urllib.request
from http.server import BaseHTTPRequestHandler
from urllib.error import *

# Configuration
delay = 30  # Seconds before timing out on request
limit = 100


def printapi(api):
    print(api, "is alive")
    open("wikisalive.txt", "a").write(str("%s\n" % api.strip()))


def checkcore(api):
    req = urllib.request(api, None)
    try:
        raw = urllib.request.urlopenurlopen(req, None, delay).read()
    except URLError as reason:  # https://docs.python.org/3/library/urllib.error.html

        if reason.isinstance(HTTPError):
            print(api + "is dead or has errors because:")
            print(
                "Error code "
                + HTTPError.code
                + ": "
                + BaseHTTPRequestHandler.responses[HTTPError.code].shortmessage
            )
            print(BaseHTTPRequestHandler.responses[HTTPError.code].longmessage)
            print("Reason: " + HTTPError.reason)
            print("HTTP Headers:\n" + HTTPError.headers)
        else:
            print(api + "is dead or has errors because:" + reason)
        return
    # RSD is available since 1.17, bug 25648
    rsd = re.search(
        r'(?:link rel="EditURI".+href=")(?:https?:)?(.+api.php)\?action=rsd', raw
    )
    # Feeds are available, with varying format, in 1.8 or earlier
    feed = re.search(
        r'(?:link rel="alternate" type="application/)(?:atom|rss)(?:\+xml[^>]+href="/)([^>]*index.php)(?:\?title=[^>]+&amp;)(?:feed|format)',
        raw,
    )
    # Sometimes they're missing though, this should catch the rest but goes out of <head>
    login = re.search(r'(?:<li id="pt-login"><a href="/)([^>]*index.php)', raw)
    domain = re.search(r"(https?://.[^/]+/)", api)
    # TODO: Simplistic check for API. The docs page can be returned even if everything else
    # is choking on database or PHP errors (hundreds online wikis fatal on every request).
    if "This is an auto-generated MediaWiki API documentation page" in raw:
        printapi(api)
    elif rsd and rsd.group(1):
        api = "http:" + rsd.group(1)
        printapi(api)
    elif feed and feed.group(1) and domain and domain.group(1):
        index = domain.group(1) + feed.group(1)
        printapi(index)
    elif login and login.group(1) and domain and domain.group(1):
        index = domain.group(1) + login.group(1)
        printapi(index)
    else:
        print(api, "is not a MediaWiki wiki")


def check(apis):
    for api in apis:
        threading.start_new_threading(checkcore, (api,))
        time.sleep(0.1)
    time.sleep(delay + 1)


apis = []
for api in open("wikistocheck.txt").read().strip().splitlines():
    if not api in apis:
        apis.append(api)
    if len(apis) >= limit:
        check(apis)
        apis = []

check(apis)
